### DISCUSSION TITLE
Unexpected validation behavior

### DISCUSSION THREAD

**Discussion by aerni** [livewire/livewire#8310] [7. Bugs (archived)]
### Livewire version

v3.4.10

### Laravel version

v10.47.0

### Which PHP version are you using?

PHP 8.1

## The Issue
I’m running into real-time validation issues after upgrading from Livewire 2 to Livewire 3. 

Validating an array property in the `updated()` hook using `validateOnly()` doesn’t work as expected. Real-time validation of the property doesn’t work at all, and validation errors of other fields are reset whenever the property is updated.

This used to work in Livewire 2 but doesn't work in Livewire 3:

```php
public function updated($property)
{
    $this->validateOnly($property);
}
```

## The Workaround

This issue is only present when manually calling the `validateOnly()` method as shown above. If we use the `#[Validate]` attribute on the property instead, real-time validation works as expected and validation errors of other fields aren’t reset as well.

*Unfortunately, I'm unable to use this workaround in my project as I'm dealing with a deeply nested custom property and the `#[Validate]` attribute doesn't trigger real-time validation as expected.*

## Demo

The following demos and videos provide a quick demonstration of the issue and the workaround.

### Issue Demo

This video shows the issue that happens when using `validateOnly()` inside the `updated()` hook: https://github.com/livewire/livewire/assets/23167701/deeb3e5b-4aa0-4b54-affb-2a1e94e1c28b

You can find this demo in the Wirebox on the `/` route.

### Workaround Demo
This video shows the working example using the `#[Validate]` attribute workaround: https://github.com/livewire/livewire/assets/23167701/fa52933f-b86f-4afc-93c8-cbdeea0fc969

You can find this demo in the Wirebox on the `/working` route.

## Why this happens

I’ve tracked the issue down to the `$fullPath` being passed to the hooks. *The `$fullPath` of the `notifications` property is `notifications.0`.*

https://github.com/livewire/livewire/blob/cac2f7b28f37980653aaec99d9de55de21b2ebaa/src/Features/SupportLifecycleHooks/SupportLifecycleHooks.php#L48-L82

Comparing this to Livewire 2, we see that we used to pass the `$name` of the property to the hooks. *The `$name` of the `notifications` property is `notifications`.*

https://github.com/livewire/livewire/blob/c3e0031652dd7387f0a6c8cd605a4d0d253a7145/src/ComponentConcerns/HandlesActions.php#L57-L103

Looking at how the `#[Validate]` attribute handles this, we see that we are also passing the `$name` of the property to the `validateOnly()` method. Which is the same behavior as in the `updated()` hook in Livewire 2. This explains, why validation works as expected using the attribute, compared to calling `validateOnly()` in the `updated()` hook in Livewire 3.

https://github.com/livewire/livewire/blob/cac2f7b28f37980653aaec99d9de55de21b2ebaa/src/Features/SupportValidation/BaseValidate.php#L86-L102

## Possible solutions

The issue seems to be caused by the `$fullPath` being passed to the hooks instead of the `$name` of the property.

### Solution 1
The sledgehammer solution would be to change it back to passing the `$name` instead of `$fullPath`. But this would be a breaking change. I'm not exactly sure why this was changed compared to Livewire 2 in the first place? If it was an intentional change, it should be documented as a breaking change. 

### Solution 2
Another potential solution could be to add some additional logic to the `validateOnly()` method to convert the `$fullPath` back to the `$name`.

https://github.com/livewire/livewire/blob/cac2f7b28f37980653aaec99d9de55de21b2ebaa/src/Features/SupportValidation/HandlesValidation.php#L321

### Solution 3
There's probably a better solution to all of this that I don't see.

### Repository or Wirebox URL

https://wirebox.app/b/4zo03

### Additional information

I found issue https://github.com/livewire/livewire/discussions/3054 and https://github.com/livewire/livewire/discussions/7839 that seem to be related but don't provide nearly as much detail.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by PerryvanderMeer** [comment]
Can you please try if the problem is solved with https://github.com/livewire/livewire/pull/8262?
It's not yet tagged, but will be in the next release.

Edit: My bad, I thought it was about a form, but it's not.

**Comment by aerni** [comment]
Unfortunately, this PR doesn't solve the issue.

**Comment by aerni** [comment]
I've created PR https://github.com/livewire/livewire/pull/8320 with a couple of tests regarding this issue.
